gh-95273: Improve sqlite3.complete_statement docs (#95840)
python/cpython: Doc/library/sqlite3.rst

.. function:: complete_statement(statement)

>    Return ``True`` if the string *statement* appears to contain one or more
>    complete SQL statements terminated by semicolons.
>    No syntactic verification or parsing of any kind is performed.

Reviewer: Reading the new version (before reading the old), it was very confusing to me what it _does_ do, if it doesn't do any syntactic verification or parsing of any kind, whereas reading the old version it was then much more clear. I suggest either reverting to something more similar to the latter, or rewriting this in a different, more explicit way.

One possibility combining the two:

```suggestion
   Return ``True`` if the string *statement* appears to contain
   one or more complete SQL statements.
   No syntactic verification or parsing of any kind is performed,
   other than checking that there are no unclosed string literals
   and the statement is terminated by a semicolon.
```

Author: True, that's better.

   This function may be useful during command-line input
   to determine if the entered text seems to form a complete SQL statement,

>    or if additional input is needed before :meth:`executing <Cursor.execute>`

Reviewer: It currently displays rather strangely because `executing` is rendered as a literal, which is not what is intended (as it is not). To work around this, I suggest (with `()`)

```suggestion
   or if additional input is needed before :meth:`~Cursor.execute`-ing
```

or (without `()`)

```suggestion
   or if additional input is needed before :meth:`execute <Cursor.execute>`-ing
```

.. function:: complete_statement(statement)

   Return ``True`` if the string *statement* appears to contain
   one or more complete SQL statements.
   No syntactic verification or parsing of any kind is performed,
   other than checking that there are no unclosed string literals

>    and the statment is terminated by a semicolon.

Reviewer 2: ```suggestion
   and the statement is terminated by a semicolon.
```

Reviewer: My fault, I assume, sorry

   This function may be useful during command-line input
   to determine if the entered text seems to form a complete SQL statement,
   or if additional input is needed before calling :meth:`~Cursor.execute`.

>    See :source:`Lib/sqlite3/__main__.py` for example use.

Reviewer 2: The linked example doesn't seem particularly useful, since it's not too obvious
* what the input is
* why the function returns `True` when it's not complete

The fact that the function doesn't have a docstring doesn't help either.

---

Since `complete_statement` seems trivial enough (it doesn't need a DB or any complex setup before using it), what about just adding something like this (with whatever query you think would better illustrate "completeness"):
```py
>>> sqlite3.complete_statement('SELECT foo FROM bar;')
True
>>> sqlite3.complete_statement('SELECT foo')
False
```

Author: > The linked example doesn't seem particularly useful, since it's not too obvious
> * what the input is
> * why the function returns `True` when it's not complete
>
> The fact that the function doesn't have a docstring doesn't help either.

Good points; this can easily be improved by telling the user which `__main__.py` function to look for in the docs, and of course by adding docstrings to `__main__.py`.

A trivial example could work well for the 3.11 and 3.10 docs, which don't have the CLI.

Reviewer: @ezio-melotti makes a great point; IMO, the minimal example would be good to include regardless, as it clearly and succinctly illustrates the purpose of the function, and when it is expected to be used. In the 3.12 docs, you could mention `__main__` (perhaps with a manual permalink to a specific line) as an additional, more complete example beyond that, but I certainly don't think it replaces the former in terms of function.

Reviewer 2: > you could mention __main__ (perhaps with a manual permalink to a specific line)

FWIW ``:source:`Lib/sqlite3/__main__.py#L39` `` seems to work as a link, but it's not "perma" and the anchor is rendered too. 
``:source:`Lib/sqlite3/__main__.py <https://github.com/python/cpython/blob/3646f6cd880f8f91e189a2fe44a687798aa1fef1/Lib/sqlite3/__main__.py#L39>` `` appends the full link to `https://github.com/python/cpython/tree/main/...` so it doesn't work.  If you want to add a permalink, you'll have to use regular rst links (otherwise you can just use the minimal example).

Author: After pondering about this for a day, I'm not so sure it is a good idea to use the CLI as an example. I'm not totally opposed to it, though. Do we even need the trivial example? There are more complex functions in the sqlite3 API without examples. Perhaps a parameter listing would be clearer: we could then specify two `:returns:` directives.

Author: > Do we even need the trivial example?

Seeing now that CAM also suggest the trivial example. I'm fine with that; let's add it.

Author: I tried adding some explanatory comments to the CLI to make it better suited as an example. However, it is hard for me to see the comments actually serve their purpose, since I wrote the CLI in the first place; everything in there is obvious to me :)

>    See :source:`Lib/sqlite3/__main__.py` for example use.

Reviewer: Implemented as a suggestion:

```suggestion
   For example::

       >>> sqlite3.complete_statement('SELECT foo FROM bar;')
       True
       >>> sqlite3.complete_statement('SELECT foo')
       False

   See :source:`Lib/sqlite3/__main__.py` for a more complete real-world use.
```

.. function:: complete_statement(statement)

   Return ``True`` if the string *statement* appears to contain
   one or more complete SQL statements.
   No syntactic verification or parsing of any kind is performed,
   other than checking that there are no unclosed string literals
   and the statement is terminated by a semicolon.

   This function may be useful during command-line input
   to determine if the entered text seems to form a complete SQL statement,
   or if additional input is needed before calling :meth:`~Cursor.execute`.

   For example::

>        >>> sqlite3.complete_statement('SELECT foo FROM bar;')
>        True
>        >>> sqlite3.complete_statement('SELECT foo')

Reviewer: My mistake for not noticing this when making the original suggestion, but for consistency with the other examples:

```suggestion
       >>> sqlite3.complete_statement("SELECT foo FROM bar;")
       True
       >>> sqlite3.complete_statement("SELECT foo")
```

Author: Consistency FTW!

       False

>    See ``runsource()`` in :source:`Lib/sqlite3/__main__.py` for real-world use.

Reviewer: Or, to be semantically perfectionist

```suggestion
   See :func:`~!sqlite.__main__.runsource` in :source:`Lib/sqlite3/__main__.py`
   for real-world use.
```

Reviewer: Or, shorter and slightly less perfectionist:

```suggestion
   See :func:`runsource` in :source:`Lib/sqlite3/__main__.py` for real-world use.
```

Author: I learn new Sphinx hacks from you every day 😆
